File: src/services/listings.ts
import { API_BASE_URL } from './config'
import type { Paper, CodeUrlInfo } from '@/types'
import { TransformBackendPaper } from '@/types'

const LISTINGS_API_BASE = `${API_BASE_URL}/api/listings`

export const listingsAPI = {
  async fetchNewListings(subject: string = 'cs'): Promise<{
    success: boolean
    date: string
    subject: string
    new_count: number
    cross_count: number
    replacement_count: number
    total_count: number
    error?: string
  }> {
    const params = new URLSearchParams({ subject })
    const response = await fetch(`${LISTINGS_API_BASE}/fetch?${params}`, {
      method: 'POST'
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return response.json()
  },

  async getListingsIndexes(subject: string = 'cs'): Promise<{
    indexes: Array<{
      date: string
      new_count: number
      cross_count: number
      replacement_count: number
      fetched_at: string
    }>
    subject: string
  }> {
    const params = new URLSearchParams({ subject })
    const response = await fetch(`${LISTINGS_API_BASE}/indexes?${params}`)

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return response.json()
  },

  async getListingsByDate(
    date: string,
    listingType: 'new' | 'cross' | 'replacement' = 'new',
    start: number = 0,
    maxResults: number = 50,
    subject: string = 'cs'
  ): Promise<{
    papers: Paper[]
    total: number
    date: string
    listing_type: string
    start: number
    max_results: number
    subject: string
  }> {
    const params = new URLSearchParams({
      listing_type: listingType,
      start: start.toString(),
      max_results: maxResults.toString(),
      subject
    })

    const response = await fetch(`${LISTINGS_API_BASE}/${date}?${params}`)

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return {
      ...data,
      papers: (data.papers || []).map(TransformBackendPaper)
    }
  },

  async getLatestListings(date?: string, subject: string = 'cs'): Promise<{
    date: string
    subject: string
    new: Paper[]
    cross: Paper[]
    replacement: Paper[]
    auto_refreshed: boolean
    error?: string
  }> {
    const params = new URLSearchParams()
    if (date) {
      params.append('date', date)
    }
    params.append('subject', subject)

    const url = `${LISTINGS_API_BASE}/new?${params}`

    const response = await fetch(url)

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return {
      ...data,
      new: (data.new || []).map(TransformBackendPaper),
      cross: (data.cross || []).map(TransformBackendPaper),
      replacement: (data.replacement || []).map(TransformBackendPaper)
    }
  },

  async checkPapersWithCode(paperIds: string[]): Promise<Record<string, boolean>> {
    if (!paperIds || paperIds.length === 0) {
      return {}
    }
    
    const response = await fetch(`${LISTINGS_API_BASE}/codes/check`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ paper_ids: paperIds })
    })
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }
    
    const data = await response.json()
    return data.codes || {}
  },

  async getCodesForPapers(paperIds: string[]): Promise<Record<string, CodeUrlInfo | null>> {
    if (!paperIds || paperIds.length === 0) {
      return {}
    }
    
    const response = await fetch(`${LISTINGS_API_BASE}/codes/batch`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ paper_ids: paperIds })
    })
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }
    
    const data = await response.json()
    const result: Record<string, CodeUrlInfo | null> = {}
    
    for (const [paperId, code] of Object.entries(data)) {
      if (code) {
        result[paperId] = {
          id: (code as Record<string, unknown>).id as string,
          paperId: (code as Record<string, unknown>).paper_id as string,
          url: (code as Record<string, unknown>).url as string,
          platform: (code as Record<string, unknown>).platform as string,
          owner: (code as Record<string, unknown>).owner as string | undefined,
          repo: (code as Record<string, unknown>).repo as string | undefined,
          isOfficial: (code as Record<string, unknown>).is_official as boolean,
          stars: (code as Record<string, unknown>).stars as number | undefined,
          language: (code as Record<string, unknown>).language as string | undefined,
          fetchedAt: (code as Record<string, unknown>).fetched_at as string | undefined,
        }
      } else {
        result[paperId] = null
      }
    }
    
    return result
  },

  async getPapersWithCode(date: string): Promise<{
    date: string
    new: Paper[]
    cross: Paper[]
    replacement: Paper[]
  }> {
    const response = await fetch(`${LISTINGS_API_BASE}/codes/papers`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ date })
    })
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }
    
    const data = await response.json()
    return {
      ...data,
      new: (data.new || []).map(TransformBackendPaper),
      cross: (data.cross || []).map(TransformBackendPaper),
      replacement: (data.replacement || []).map(TransformBackendPaper)
    }
  }
}

File: backend/app/routers/listings.py
import logging
from fastapi import APIRouter, Query
from fastapi import HTTPException
from pydantic import BaseModel
from typing import Optional, List, Dict

from app.services.listings_service import ListingsService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/listings", tags=["listings"])

_listings_service = ListingsService()


class PaperCodeCheckRequest(BaseModel):
    paper_ids: List[str]


class PaperCodeCheckResponse(BaseModel):
    codes: Dict[str, bool]


class PaperCodeBatchRequest(BaseModel):
    paper_ids: List[str]


@router.post("/fetch")
async def fetch_new_listings(
    subject: str = Query('cs', description="Subject category to fetch: 'cs', 'q-fin', or 'stat'")
):
    """
    Fetch and store arXiv new listings.

    Scrapes the arXiv new listings page (https://arxiv.org/list/{subject}/new),
    extracts paper IDs for new, cross, and replacement submissions,
    fetches paper details from arXiv API, and stores them in the database.

    Args:
        subject: Subject category (cs, q-fin, stat). Default is 'cs'.
    """
    try:
        result = await _listings_service.fetch_and_store_listings(subject)
        return result
    except Exception as e:
        logger.error(f"Error fetching new listings: {e}")
        return {
            "success": False,
            "error": str(e),
            "subject": subject,
            "new_count": 0,
            "cross_count": 0,
            "replacement_count": 0,
            "total_count": 0,
        }


@router.get("/indexes")
async def get_listings_indexes(
    subject: str = Query('cs', description="Subject category (cs, q-fin, stat)")
):
    """
    Get all listings date indexes for a specific subject.

    Returns a list of all dates that have listings data stored,
    with counts for each type of submission.

    Args:
        subject: Subject category (cs, q-fin, stat). Default is 'cs'.
    """
    try:
        indexes = _listings_service.get_listings_indexes(subject)
        return {"indexes": indexes, "subject": subject}
    except Exception as e:
        logger.error(f"Error getting listings indexes: {e}")
        return {"indexes": [], "subject": subject}


@router.get("/new")
async def get_latest_listings(
    date: Optional[str] = Query(None, description="Date in YYYY-MM-DD format. If not specified, get latest with auto-refresh"),
    subject: str = Query('cs', description="Subject category to fetch: 'cs', 'q-fin', or 'stat'")
):
    """
    Get the latest day's listings for all three types with auto-refresh.

    If date is specified, return papers for that date without auto-refresh.
    If date is not specified, return latest papers with auto-refresh.

    Args:
        subject: Subject category (cs, q-fin, stat). Default is 'cs'.

    Returns:
        {
            "date": "2026-04-13",
            "subject": "cs",
            "new": [...],
            "cross": [...],
            "replacement": [...],
            "auto_refreshed": true
        }
    """
    try:
        result = await _listings_service.get_latest_listings(date, subject)
        return result
    except Exception as e:
        logger.error(f"Error getting latest listings: {e}")
        return {
            "date": "",
            "subject": subject,
            "new": [],
            "cross": [],
            "replacement": [],
            "auto_refreshed": False,
            "error": str(e),
        }


@router.get("/{date}")
async def get_listings_by_date(
    date: str,
    listing_type: str = Query("new", description="Type of listings: 'new', 'cross', or 'replacement'"),
    start: int = Query(0, ge=0, description="Start index for pagination"),
    max_results: int = Query(50, ge=1, le=5000, description="Maximum papers to return"),
    subject: str = Query('cs', description="Subject category (cs, q-fin, stat)")
):
    """
    Get listings by date and type.

    Args:
        date: Date in YYYY-MM-DD format
        listing_type: 'new', 'cross', or 'replacement'
        subject: Subject category (cs, q-fin, stat). Default is 'cs'.
    """
    try:
        result = _listings_service.get_listings_by_date(
            date=date,
            listing_type=listing_type,
            start=start,
            max_results=max_results,
            subject=subject
        )
        return result
    except Exception as e:
        logger.error(f"Error getting listings: {e}")
        return {
            "papers": [],
            "total": 0,
            "date": date,
            "listing_type": listing_type,
            "start": start,
            "max_results": max_results,
            "subject": subject,
            "error": str(e),
        }


@router.post("/codes/check", response_model=PaperCodeCheckResponse)
async def check_papers_with_code(request: PaperCodeCheckRequest):
    """
    Check which papers have code repositories.
    
    Returns a dictionary mapping paper_id to boolean (True if has code).
    """
    try:
        if not request.paper_ids:
            return PaperCodeCheckResponse(codes={})
        result = _listings_service.check_papers_with_code(request.paper_ids)
        return PaperCodeCheckResponse(codes=result)
    except Exception as e:
        logger.error(f"Error checking papers with code: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/codes/batch")
async def get_codes_for_papers(request: PaperCodeBatchRequest):
    """
    Get code repositories for multiple papers.
    
    Returns a dictionary mapping paper_id to code repository info (or null if not found).
    """
    try:
        if not request.paper_ids:
            return {}
        result = _listings_service.get_codes_for_papers(request.paper_ids)
        return result
    except Exception as e:
        logger.error(f"Error getting codes for papers: {e}")
        raise HTTPException(status_code=500, detail=str(e))


class PapersWithCodeRequest(BaseModel):
    date: str


@router.post("/codes/papers")
async def get_papers_with_code(request: PapersWithCodeRequest):
    """
    Get papers with code repositories for a specific date.
    
    Returns papers that have code repository links, grouped by submission type.
    
    Returns:
        {
            "date": "2026-04-22",
            "new": [...],
            "cross": [...],
            "replacement": [...]
        }
    """
    try:
        result = _listings_service.get_papers_with_code_by_date(request.date)
        return result
    except Exception as e:
        logger.error(f"Error getting papers with code: {e}")
        raise HTTPException(status_code=500, detail=str(e))
